File: src/main/java/dev/nytweetdeck/list/ListDirectoryParser.java
package dev.nytweetdeck.list;

import dev.nytweetdeck.list.ListDirectoryPage.ListOption;
import dev.nytweetdeck.xapi.http.XApiHttpException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.stereotype.Component;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

@Component
public class ListDirectoryParser {

    private final ObjectMapper objectMapper;

    public ListDirectoryParser(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public ListDirectoryPage parse(String body, String source) {
        try {
            var root = objectMapper.readTree(body);
            var lists = new LinkedHashMap<String, ListOption>();
            var cursor = new String[1];
            visit(root, source, lists, cursor);
            return new ListDirectoryPage(new ArrayList<>(lists.values()), cursor[0]);
        } catch (JacksonException exception) {
            throw new XApiHttpException("Xリスト一覧を解析できません。", exception);
        }
    }

    private static void visit(
            JsonNode node, String source, Map<String, ListOption> lists, String[] cursor) {
        if (node == null || node.isNull()) {
            return;
        }
        if (node.isArray()) {
            for (var child : node) {
                visit(child, source, lists, cursor);
            }
            return;
        }
        if (!node.isObject()) {
            return;
        }
        var cursorType = text(node, "cursorType");
        if ("Bottom".equalsIgnoreCase(cursorType)) {
            cursor[0] = text(node, "value");
        }
        var list = "TimelineTwitterList".equals(text(node, "__typename"))
                ? object(node, "list")
                : null;
        if (list != null && text(list, "id_str") != null && text(list, "name") != null) {
            var owner = list.path("user_results").path("result");
            var ownerCore = owner.path("core");
            lists.putIfAbsent(
                    text(list, "id_str"),
                    new ListOption(
                            text(list, "id_str"),
                            text(list, "name"),
                            text(list, "description"),
                            text(ownerCore, "name"),
                            text(ownerCore, "screen_name"),
                            number(list, "member_count"),
                            number(list, "subscriber_count"),
                            source));
            return;
        }
        for (Map.Entry<String, JsonNode> property : node.properties()) {
            visit(property.getValue(), source, lists, cursor);
        }
    }

    private static String text(JsonNode node, String field) {
        var value = node == null ? null : node.get(field);
        return value == null || value.isNull() ? null : value.asString(null);
    }

    private static JsonNode object(JsonNode node, String field) {
        var value = node == null ? null : node.get(field);
        return value != null && value.isObject() ? value : null;
    }

    private static long number(JsonNode node, String field) {
        var value = node == null ? null : node.get(field);
        return value == null ? 0 : value.asLong(0);
    }
}

File: android/app/src/main/java/dev/nytweetdeck/android/xapi/ListDirectoryParser.kt
package dev.nytweetdeck.android.xapi

import dev.nytweetdeck.android.model.ListDirectoryPage
import dev.nytweetdeck.android.model.ListOption
import java.util.LinkedHashMap
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.longOrNull

/** X Web APIのリスト一覧応答をカラム候補へ正規化する。 */
class ListDirectoryParser(
    private val json: Json = Json,
) {
    fun parse(
        body: String,
        source: String,
    ): ListDirectoryPage = try {
        val root = json.parseToJsonElement(body)
        val lists = LinkedHashMap<String, ListOption>()
        val cursor = CursorHolder()
        visit(root, source, lists, cursor)
        ListDirectoryPage(lists.values.toList(), cursor.value)
    } catch (exception: XApiException) {
        throw exception
    } catch (exception: Exception) {
        throw XApiException("Xリスト一覧を解析できません。", cause = exception)
    }

    private fun visit(
        node: JsonElement?,
        source: String,
        lists: MutableMap<String, ListOption>,
        cursor: CursorHolder,
    ) {
        when (node) {
            is JsonArray -> node.forEach { visit(it, source, lists, cursor) }
            is JsonObject -> {
                findCursor(node, cursor)
                parseList(node, source)?.let { list ->
                    lists.putIfAbsent(list.id, list)
                    return
                }
                node.values.forEach { visit(it, source, lists, cursor) }
            }

            else -> Unit
        }
    }

    private fun findCursor(
        node: JsonObject,
        cursor: CursorHolder,
    ) {
        if (node.text("cursorType").equals("Bottom", ignoreCase = true)) {
            cursor.value = node.text("value")
        }
    }

    private fun parseList(
        node: JsonObject,
        source: String,
    ): ListOption? {
        if (node.text("__typename") != "TimelineTwitterList") {
            return null
        }
        val list = node.objectValue("list") ?: return null
        val id = list.text("id_str") ?: return null
        val name = list.text("name") ?: return null
        val ownerCore = list.objectValue("user_results")
            .objectValue("result")
            .objectValue("core")
        return ListOption(
            id = id,
            name = name,
            description = list.text("description"),
            ownerName = ownerCore.text("name"),
            ownerUsername = ownerCore.text("screen_name"),
            memberCount = list.longValue("member_count"),
            subscriberCount = list.longValue("subscriber_count"),
            source = source,
        )
    }

    private fun JsonObject?.text(name: String): String? =
        (this?.get(name) as? JsonPrimitive)?.contentOrNull

    private fun JsonObject?.objectValue(name: String): JsonObject? =
        this?.get(name) as? JsonObject

    private fun JsonObject?.longValue(name: String): Long =
        (this?.get(name) as? JsonPrimitive)?.longOrNull ?: 0L

    private class CursorHolder {
        var value: String? = null
    }
}
